src/lib.rs

```diff
@@ -667,6 +669,34 @@ decl_module! {
 
             Ok(())
         }
+
+        #[weight = FunctionOf(
+            |args: (&T::AccountId, &u8, &Timepoint<T::BlockNumber>)| 500_000 * (*args.1 as u32),
+            DispatchClass::Normal,
+            true
+        )]
+		fn clear(origin, multi_account_id: T::AccountId, max_cancellations: u8, timepoint: Timepoint<T::BlockNumber>) -> DispatchResult {
+			let mut succeeded = 0;
+			let who = ensure_signed(origin)?;
+			let mut cancelled_multisigs = Vec::with_capacity(max_cancellations as usize);
+			let active_multisigs = <Multisigs<T>>::iter_prefix(&who);
+			for m_sig in active_multisigs.take(max_cancellations as usize){
+				ensure!(who == multi_account_id || who == m_sig.depositor, Error::<T>::NotOwner);
+
+				let _ = T::Currency::unreserve(&m_sig.depositor, m_sig.deposit);
+				cancelled_multisigs.push(m_sig);
+				succeeded += 1;
+			}
+
+			<Multisigs<T>>::remove_prefix(&multi_account_id);
```

Comment:
We cannot do this since we might not have checked all the multisigs in the loop above. Please use a second loop for removing the multisigs individually.

Reply:
We had this code in the earlier commits, 
```rust
<Multisigs<T>>::remove(&multi_account_id, call_hash);
```
but individually removing multisigs requires a `call_hash` (which we don't have).

Reply:
Reading at what take might do:
https://github.com/paritytech/substrate/blob/d2627c48acbdf08f411f2a998876c6acb97ce92a/frame/support/src/storage/mod.rs#L284
We might not even need the `remove_prefix` right? Or I am missing something. I think testing it out with some tests will get us out of any doubt.

Reply:
@mikiquantum I think `iter_prefix_values` should do the trick. https://github.com/paritytech/substrate/blob/d2627c48acbdf08f411f2a998876c6acb97ce92a/frame/support/src/storage/mod.rs#L314

This would return both the `multi_sig` and the `value` (call_hash?) for a certain `multi_account_id` then we could iterate and call on `remove(multi_sig, value)` up to `max_cancellation`.
However, i'm getting a compilation error:
```rust
164 | decl_storage! {
    | ------------- function or associated item `iter_prefix_values` not found for this
...
683 |             let active_multisigs = <Multisigs<T>>::iter_prefix_values(&multi_account_id);
```
not sure why the compiler can not find it when `iter_prefix_values` is surely there in the functions

Reply:
`iter_prefix_value` is just a rename to `iter_prefix` and has just been added [recently](https://github.com/paritytech/substrate/pull/5504/files), so it doesn't help here.

Reply:
Is not `take` doing the trick?

Reply:
`take` needs 2 parameters, which I think is the `call_hash` which we don't have.
Multisigs is a [StorageDoubleMap](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html) which has 2 keys, (multi_account_id, call_hash) and the value is the `multi_sig`.
Somehow we need a way to get all the `call_hash` on this Multisigs that under the first key `multi_account_id`.

Reply:
If my assumptions is correct, `Multisigs` is storing the values like this:
```
multi_id1 => call_hash1 => msig1
	     call_hash2 => msig2
	     call_hash3 => msig3

multi_id2 => call_hash7 => msig101
          => call_hash8 => msig102
```

Reply:
That's correct @ivanceras . Since you are iterating overt he multisigs, you should be able to get the call_hash. I am not 100% sure how to actually do that, could you ask in https://riot.im/app/#/room/#substrate-technical:matrix.org ?

Reply:
It looks like we need [`drain_prefix`](https://github.com/paritytech/substrate/blob/d2627c48acbdf08f411f2a998876c6acb97ce92a/frame/support/src/storage/generator/double_map.rs#L386).
We can iterate over up to the `max_cancellation` while the values is being removed from the storage. The big problem is that, it requires we need to use the new version of `substrate`, and there are a lot of codebase that needs to be migrated for this new version.

Reply:
@philipstanislaus would it be a solution to keep track of how many has been cancelled and how many is left, and then if the count is zero, that's when we call `remove_prefix`.
```rust
if cancelled_count == 0{
   <Multisigs<T>>::remove_prefix(&multi_account_id);
}
```
The downside is this adds a new entry to the storage.